File: src/003. Largest prime factor/003.go
package main

import "fmt"

func main() {	
	fmt.Println(maxPrimeFactor(600851475143))
}

func maxPrimeFactor(num int) int {
	// 2 is the only even prime
	for num % 2 == 0 {
		num /= 2
	}

	max := 2
	for factor := 3; num > 1; factor += 2 {
		for num % factor == 0 {
			num /= factor
			max = factor
		}
	}
	return max
}

File: src/003. Largest prime factor/003.cpp
#include <iostream>

long long max_prime_factor(long long num);

int main(void) {  
    std::cout << max_prime_factor(600851475143LL) << std::endl;
    return 0;
}

long long max_prime_factor(long long num) {
    // 2 is the only even prime number
    while (num % 2 == 0) {
        num /= 2;
    }

    long long max = 2;
    long long factor = 3;
    while (num > 1) {
        while (num % factor == 0) {
            num /= factor;
            max = factor;
        }
        factor += 2;
    }
    return max;
}
